crates/accelerate/src/two_qubit_decompose.rs

```diff
@@ -0,0 +1,181 @@
+                                           [c0, c0, c1im, c1],
+                                           [c0, c0, c1im, -c1],
+                                           [c1, -c1im, c0, c0]];
+    let _B_nonnormalized_dagger = Scale(c64 {re: 0.5, im: 0.0}) * _B_nonnormalized.adjoint();
+    return _B_nonnormalized_dagger * U * _B_nonnormalized;
+}
+
+// faer::c64 and num_complex::Complex<f64> are both structs
+// holding two f64's. But several functions are not defined for
+// c64. So we implement them here. These things should be contribute
+// upstream.
+
+pub trait PowF {
+    fn powf(self, pow : f64) -> c64;
+}
+
+impl PowF for c64 {
+    fn powf(self, pow : f64) -> c64 {
+        return c64::from(Complex::<f64>::from(self).powf(pow));
+    }
+}
+
+pub trait Arg {
+    fn arg(self) -> f64;
+}
+
+impl Arg for c64 {
+    fn arg(self) -> f64 {
+        let c = Complex::<f64>::from(self);
+        return c.arg();
+    }
+}
+
+// FIXME: full of ineffciencies
+fn __weyl_coordinates(unitary : MatRef<c64>) -> (f64, f64, f64) {
+    let pi = PI;
+    let pi2 = PI / 2.0;
+    let pi4 = PI / 4.0;
+    let uscaled = Scale(c1 / unitary.determinant().powf(0.25)) * unitary;
+    let uup = transform_from_magic_basis(uscaled);
+    let uad = uup.transpose();
+    let mut d : Vec<c64> = (&uad * &uup).eigenvalues();
+    let mut darg : Vec<_> = d.iter().map(|x| - x.arg() / 2.0).collect();
+    darg[3] = -darg[0] - darg[1] - darg[2];
+    let mut cs = vec![0.0; 3];
+    for i in 0..3 {
+        cs[i] = utils::modulo((darg[i] + darg[3]) / 2.0, 2.0 * pi);
+    }
+    let mut cstemp : Vec<f64> = cs.iter().map(|x| utils::modulo(*x, pi2)).collect();
+    for i in 0..3 {
+        cstemp[i] = cstemp[i].min(pi2 - cstemp[i]);
+    }
+    let mut order = utils::argsort(&cstemp);
+    (order[0], order[1], order[2]) = (order[1], order[2], order[0]);
+    (cs[0], cs[1], cs[2]) = (cs[order[0]], cs[order[1]], cs[order[2]]);
+
+    // Flip into Weyl chamber
+    let pi32 = 3.0 * pi2;
+    if cs[0] > pi2 {
+        cs[0] -= pi32;
+    }
+    if cs[1] > pi2 {
+        cs[1] -= pi32;
+    }
+    let mut conjs = 0;
+    if cs[0] > pi4 {
+        cs[0] = pi2 - cs[0];
+        conjs += 1;
+    }
+    if cs[1] > pi4 {
+        cs[1] = pi2 - cs[1];
+        conjs += 1;
+    }
+    if cs[2] > pi2 {
+        cs[2] -= pi32;
+    }
+    if conjs == 1 {
+        cs[2] = pi2 - cs[2];
+    }
+    if cs[2] > pi4 {
+        cs[2] -= pi2;
+    }
+    return (cs[1], cs[0], cs[2]);
+}
+
+
+// For debugging. We can remove this later
+#[pyfunction]
+#[pyo3(text_signature = "(unitary, /")]
+pub fn _weyl_coordinates(unitary : PyReadonlyArray2<Complex<f64>>) -> (f64, f64, f64) {
+    let u = unitary.as_array().into_faer_complex();
+    return __weyl_coordinates(u);
+}
+
+#[pyfunction]
+#[pyo3(text_signature = "(basis_b, basis_fidelity, unitary, /")]
+pub fn _num_basis_gates(basis_b : f64, basis_fidelity : f64, unitary : PyReadonlyArray2<Complex<f64>>) -> usize {
```

Comment:
One thing we can do here is call `blocks_to_matrix` (https://github.com/Qiskit/qiskit/blob/main/crates/accelerate/src/convert_2q_block_matrix.rs#L25 ) directly from rust. The output of `blocks_to_matrix` (via a layer of indirection) gets wrapped in a `UnitaryGate` and then calls `num_basis_gates` here: https://github.com/Qiskit/qiskit/blob/main/qiskit/transpiler/passes/optimization/consolidate_blocks.py#L134  to evaluate one of the conditions on whether synthesis is worth it or not (and if it is do the gate consolidation). So we could avoid some python call overhead by just returning this optionally directly from rust.

Although we probably can save this for a follow up to this PR.

Reply:
I can try to put it in here.